### PR TITLE
feat: allow specifying an expected object size for resumable operations.

### LINKED ISSUE #2511: allow specifying the content length for resumable uploads
**Is your feature request related to a problem? Please describe.**
The resumable upload API allows [specifying a final length of the object](https://cloud.google.com/storage/docs/json_api/v1/parameters#xuploadcontentlength) if known in a HTTP header. There is no way to pass this value into the Java GCS API for resumable uploads, though.

**Describe the solution you'd like**
A `Storage.writer` method overload that takes the final content length.

### PR DESCRIPTION
**Pull request by BenWhitehead**
Update resumable upload failure detection to be more specific about classifying a request as SCENARIO_5

Fixes #2511

### CODE REVIEW

File: google-cloud-storage/src/main/java/com/google/cloud/storage/GapicBidiUnbufferedWritableByteChannel.java [hunk 1]
```diff
@@ -345,10 +346,17 @@ public void onNext(BidiWriteObjectResponse value) {
...
-        clientDetectedError(
-            ResumableSessionFailureScenario.SCENARIO_5.toStorageException(
-                ImmutableList.of(lastWrittenRequest), null, context, oore));
-      } else if (t instanceof ApiException) {
+        ErrorDetails ed = oore.getErrorDetails();
+        if (!(ed != null
+            && ed.getErrorInfo() != null
+            && ed.getErrorInfo().getReason().equals("GRPC_MISMATCHED_UPLOAD_SIZE"))) {
+          clientDetectedError(
+              ResumableSessionFailureScenario.SCENARIO_5.toStorageException(
```
Comment (sydney-munro) [inline google-cloud-storage/src/main/java/com/google/cloud/storage/GapicBidiUnbufferedWritableByteChannel.java:354]: can we get some comments around what Scenario 5 is to make it easier to read?

File: google-cloud-storage/src/main/java/com/google/cloud/storage/GrpcStorageOptions.java [hunk 2]
```diff
@@ -279,7 +279,8 @@ private Tuple<StorageSettings, Opts<UserProject>> resolveSettingsAndOpts() throw
         InstantiatingGrpcChannelProvider.newBuilder()
             .setEndpoint(endpoint)
             .setAllowNonDefaultServiceAccount(true)
-            .setAttemptDirectPath(attemptDirectPath);
+            .setAttemptDirectPath(attemptDirectPath)
+            .setMaxInboundMetadataSize(50 * 1024);
```
Comment (sydney-munro) [inline google-cloud-storage/src/main/java/com/google/cloud/storage/GrpcStorageOptions.java:283]: where does this number come from?

File: google-cloud-storage/src/main/java/com/google/cloud/storage/GapicBidiUnbufferedWritableByteChannel.java (hunk 1 above)
Comment (sydney-munro) [reply google-cloud-storage/src/main/java/com/google/cloud/storage/GapicBidiUnbufferedWritableByteChannel.java:354]: oh i see this was there before as well. my apologies. that can be a followup.

**Review by sydney-munro — APPROVED** [review]

File: google-cloud-storage/src/main/java/com/google/cloud/storage/GrpcStorageOptions.java (hunk 2 above)
Comment (BenWhitehead) [reply google-cloud-storage/src/main/java/com/google/cloud/storage/GrpcStorageOptions.java:283]: woops, this we left over from some test failure debugging. I'll remove it.

File: google-cloud-storage/src/main/java/com/google/cloud/storage/GapicBidiUnbufferedWritableByteChannel.java (hunk 1 above)
Comment (BenWhitehead) [reply google-cloud-storage/src/main/java/com/google/cloud/storage/GapicBidiUnbufferedWritableByteChannel.java:354]: ResumableSessionFailureScenario is an enum that is shared by several of our upload implementations and defines the types of failure scenarios that can happen when processing an upload request.

SCENARIO_5 is client side data loss detected https://github.com/googleapis/java-storage/blob/380057bd5c6f74158e55854dc159382703ea328f/google-cloud-storage/src/main/java/com/google/cloud/storage/ResumableSessionFailureScenario.java#L72-L75

**Pull request merged**